File: generate_item/generate_item/report/requested_items_to_be_received/requested_items_to_be_received.py
# Copyright (c) 2026, Finbyz and contributors
# For license information, please see license.txt

from frappe import _
import frappe
from frappe.utils import flt, getdate, nowdate





def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {"label": "Purchase Order", "fieldname": "name", "fieldtype": "Link", "options": "Purchase Order", "width": 160},
        {"label": "Batch No Ref", "fieldname": "custom_batch_no", "fieldtype": "Link", "options": "Batch", "width": 160},
        {"label": "Supplier", "fieldname": "supplier", "fieldtype": "Link", "options": "Supplier", "width": 180},
        
        {"label": "Posting Date", "fieldname": "transaction_date", "fieldtype": "Date", "width": 110},
        {"label": "Schedule Date", "fieldname": "schedule_date", "fieldtype": "Date", "width": 110},
        {"label": "Status", "fieldname": "status", "width": 120},
        {"label": "Item ID", "fieldname": "item_id", "width": 220},
        {"label": "Item Code", "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 140},
        {"label": "Item Name", "fieldname": "item_name", "width": 220},
        {"label": "Description", "fieldname": "description", "width": 250},
        
        {"label": "PO Qty", "fieldname": "po_qty", "fieldtype": "Float", "width": 110},
        {"label": "Received Qty", "fieldname": "received_qty", "fieldtype": "Float", "width": 120},
        {"label": "Pending Receipt Qty", "fieldname": "pending_qty", "fieldtype": "Float", "width": 140},
        
        {"label": "Drawing Number", "fieldname": "custom_drawing_no", "width": 220},
        {"label": "Drawing Rev No", "fieldname": "custom_drawing_rev_no", "width": 220},
        
        {"label": "Created By", "fieldname": "created_by", "fieldtype": "Link", "options": "User", "width": 140},
        
        
        {"label": "Pattern Drawing No", "fieldname": "custom_pattern_drawing_no", "width": 220},
		{"label": "Pattern Drawing Rev No", "fieldname": "custom_pattern_drawing_rev_no", "width": 220},
		{"label": "Purchase Specification No", "fieldname": "custom_purchase_specification_no", "width": 220},
		{"label": "Purchase Specification Rev No", "fieldname": "custom_purchase_specification_rev_no", "width": 220},
  

        {"label": "Age (Days)", "fieldname": "age", "fieldtype": "Int", "width": 90},
		{"label": "0-30", "fieldname": "range_0_30", "fieldtype": "Float", "width": 100},
		{"label": "31-60", "fieldname": "range_31_60", "fieldtype": "Float", "width": 100},
		{"label": "61-90", "fieldname": "range_61_90", "fieldtype": "Float", "width": 100},
		{"label": "91-120", "fieldname": "range_91_120", "fieldtype": "Float", "width": 100},
		{"label": "121+", "fieldname": "range_121_above", "fieldtype": "Float", "width": 100},
  
        {"label": "Warehouse", "fieldname": "warehouse", "fieldtype": "Link", "options": "Warehouse", "width": 160},
        {"label": "Company", "fieldname": "company", "width": 220},
    ]



def get_data(filters):
    conditions = []
    values = {}

    if filters.get("company"):
        conditions.append("po.company = %(company)s")
        values["company"] = filters["company"]

    if filters.get("supplier"):
        conditions.append("po.supplier = %(supplier)s")
        values["supplier"] = filters["supplier"]

    # filter by created by (multi or single)
    created_by = filters.get("created_by")
    if created_by:
        if isinstance(created_by, str):
            created_by_list = [u.strip() for u in created_by.replace(",", "\n").split("\n") if u.strip()]
        else:
            created_by_list = created_by

        if created_by_list:
            conditions.append("po.owner IN %(created_by)s")
            values["created_by"] = tuple(created_by_list)

    if filters.get("from_date"):
        conditions.append("po.transaction_date >= %(from_date)s")
        values["from_date"] = filters["from_date"]

    if filters.get("to_date"):
        conditions.append("po.transaction_date <= %(to_date)s")
        values["to_date"] = filters["to_date"]

    if filters.get("status"):
        conditions.append("po.status = %(status)s")
        values["status"] = filters["status"]

    where_clause = " AND ".join(conditions)
    if where_clause:
        where_clause = " AND " + where_clause

    query = f"""
        SELECT
            po.name,
            po.owner as created_by,
            po.supplier,
            po.transaction_date,
            po.schedule_date,
            po.status,
            po.company,
            po.branch,

            poi.name AS purchase_order_item,
            poi.item_code,
            poi.item_name,
            poi.description,
            poi.custom_batch_no,
            poi.custom_drawing_no,
            poi.custom_drawing_rev_no,
            poi.qty AS po_qty,
            IFNULL(poi.received_qty, 0) AS received_qty,
            poi.rate,
            poi.warehouse,
            poi.uom,
            poi.schedule_date AS item_schedule_date,

            poi.material_request,
            poi.material_request_item,
            poi.name as item_id,
			poi.custom_pattern_drawing_no,
			poi.custom_pattern_drawing_rev_no,
			poi.custom_purchase_specification_no,
			poi.custom_purchase_specification_rev_no

        FROM `tabPurchase Order` po
        INNER JOIN `tabPurchase Order Item` poi
            ON poi.parent = po.name
        WHERE
            po.docstatus = 1
            {where_clause}
        ORDER BY
            po.transaction_date ASC, po.name
    """

    rows = frappe.db.sql(query, values, as_dict=True)

    data = []
    today = getdate(nowdate())
    for row in rows:
        pending_qty = flt(row.po_qty) - flt(row.received_qty)

        if pending_qty <= 0:
            continue
        
        age = (today - getdate(row.item_schedule_date)).days if row.item_schedule_date else 0


        row.update({
            "pending_qty": pending_qty,
            "age": age,
			"range_0_30": pending_qty if age <= 30 else 0,
			"range_31_60": pending_qty if 31 <= age <= 60 else 0,
			"range_61_90": pending_qty if 61 <= age <= 90 else 0,
			"range_91_120": pending_qty if 91 <= age <= 120 else 0,
			"range_121_above": pending_qty if age > 120 else 0,
        })

        data.append(row)

    return data



@frappe.whitelist()
def create_purchase_receipt_by_supplier(grouped_items, company, pr_series=None, branch=None):
    """
    Create Purchase Receipt grouped by supplier from Purchase Orders
    """
    import json

    if isinstance(grouped_items, str):
        grouped_items = json.loads(grouped_items)

    created_receipts = []

    try:
        for supplier, items in grouped_items.items():
            if not items:
                continue

            pr = frappe.new_doc("Purchase Receipt")
            pr.supplier = supplier
            pr.company = company
            pr.branch = branch or items[0].get("branch")

            if pr_series:
                pr.naming_series = pr_series

            pr.posting_date = nowdate()
            pr.set_posting_time = 1

            for item in items:
                pr.append("items", {
                    "item_code": item.get("item_code"),
                    "item_name": item.get("item_name"),
                    "description": item.get("description"),
                    "qty": item.get("pending_qty"),
                    "uom": item.get("uom"),
                    "rate": item.get("rate"),
                    "warehouse": item.get("warehouse"),

                    # REQUIRED LINKS
                    "purchase_order": item.get("name"),
                    "purchase_order_item": item.get("purchase_order_item"),
                    "material_request": item.get("material_request"),
                    "material_request_item": item.get("material_request_item"),

                    # CUSTOM FIELDS
                    "custom_batch_no": item.get("custom_batch_no"),
                    "custom_drawing_no": item.get("custom_drawing_no"),
                    "custom_drawing_rev_no": item.get("custom_drawing_rev_no"),
                    "custom_pattern_drawing_no":item.get("custom_pattern_drawing_no"),
                    "custom_pattern_drawing_rev_no":item.get("custom_pattern_drawing_rev_no"),
                    "custom_purchase_specification_no":item.get("custom_purchase_specification_no"),
                    "custom_purchase_specification_rev_no":item.get("custom_purchase_specification_rev_no")
                    
                    
                })

            pr.flags.ignore_permissions = True
            pr.flags.ignore_mandatory = True
            pr.save(ignore_permissions=True)

            created_receipts.append({
                "purchase_receipt": pr.name,
                "supplier": supplier,
                "items": len(items),
                "status": "Success"
            })

            frappe.db.commit()

        if created_receipts:
            frappe.msgprint(_("Successfully created {0} Purchase Receipt(s)").format(len(created_receipts)))

        return created_receipts

    except Exception:
        frappe.log_error(frappe.get_traceback(), _("Purchase Receipt Creation Error"))
        frappe.throw(_("Failed to create Purchase Receipt"))
        
@frappe.whitelist()
def get_pr_naming_series():
    meta = frappe.get_meta("Purchase Receipt")
    field = meta.get_field("naming_series")

    if not field or not field.options:
        return []

    return [opt for opt in field.options.split("\n") if opt]


@frappe.whitelist()
def get_last_purchase_history(item_codes):
    import json

    if isinstance(item_codes, str):
        item_codes = json.loads(item_codes)

    if not item_codes:
        return []

    query = """
        SELECT
            poi.item_code,
            po.name AS po_no,
            po.posting_date AS po_date,
            po.supplier,
            poi.qty,
            poi.rate
        FROM `tabPurchase Receipt Item` poi
        INNER JOIN `tabPurchase Receipt` po
            ON po.name = poi.parent
        WHERE
            poi.item_code IN %(item_codes)s
            AND po.docstatus = 1
            AND po.status NOT IN ('Closed')
        ORDER BY
            po.posting_date DESC
    """

    rows = frappe.db.sql(
        query,
        {"item_codes": tuple(item_codes)},
        as_dict=True
    )

    # Pick last (latest) purchase per item
    seen = set()
    result = []
	

    for row in rows:
        if row.item_code not in seen:
            result.append(row)
            seen.add(row.item_code)

    return result

File: generate_item/generate_item/report/requested_items_to_be_received/requested_items_to_be_received.js
// Copyright (c) 2026, Finbyz and contributors
// For license information, please see license.txt


let PO_SERIES_OPTIONS = [];
frappe.query_reports["Requested Items To Be Received"] = {

    get_datatable_options(options) {
        options.checkboxColumn = true;
        return options;
    },
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
            reqd: 1,
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        
        {
            fieldname: "created_by",
            label: __("Created By"),
            fieldtype: "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options("User", txt);
            },
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        {
            fieldname: "supplier",
            label: __("Supplier"),
            fieldtype: "Link",
            options: "Supplier",
           
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        {
            fieldname: "item_code",
            label: __("Item"),
            fieldtype: "Link",
            options: "Item",
           
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        {
            fieldname: "purchase_no",
            label: __("Purchase Order"),
            fieldtype: "Link",
            options: "Purchase Order",
           
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch",
           
            on_change: function () {
                frappe.query_report.refresh().then(() => {
                    unchecked_all_checkbox();
                });
            },
        },
        
       
      
        
        
        
    ],
    onload: function (report) {

         frappe.call({
        method: "generate_item.generate_item.report.requested_items_to_be_received.requested_items_to_be_received.get_pr_naming_series",
        callback: function (r) {
            // console.log("series-------",r)
            if (r.message) {
                
                PO_SERIES_OPTIONS = r.message;
            }
        }
    });
        report.page.add_inner_button(
            __("Past Purchase Receipt History"),
            function () {
                show_past_purchase_history();
            }
        );
    }
    
};



function show_purchase_history_dialog(data) {
    let html = `
        <div style="overflow-x:auto;">
            <table class="table table-bordered table-hover">
                <thead>
                    <tr>
                        <th>Item</th>
                        <th>PR No</th>
                        <th>PR Date</th>
                        <th>Supplier</th>
                        <th class="text-right">Qty</th>
                        <th class="text-right">Rate</th>
                    </tr>
                </thead>
                <tbody>
    `;

    data.forEach(d => {
        html += `
            <tr>
                <td style="white-space:nowrap">${d.item_code}</td>
                <td>
                    <a href="/app/purchase-receipt/${d.po_no}" target="_blank">
                        ${d.po_no}
                    </a>
                </td>
                <td>${frappe.datetime.str_to_user(d.po_date)}</td>
                <td style="max-width:260px; white-space:normal;">
                    ${d.supplier}
                </td>
                <td class="text-right">${d.qty}</td>
                <td class="text-right">${format_currency(d.rate)}</td>
            </tr>
        `;
    });

    html += `
                </tbody>
            </table>
        </div>
    `;

    let dialog = new frappe.ui.Dialog({
        title: __("Last Purchase Receipt History (Approved PR Only)"),
        size: "large",
        fields: [
            {
                fieldtype: "HTML",
                fieldname: "history_html"
            }
        ]
    });

    dialog.fields_dict.history_html.$wrapper.html(html);

    dialog.show();
}

function show_past_purchase_history() {
    const rows = frappe.query_report.get_checked_items() || [];

    if (!rows.length) {
        frappe.msgprint(__("Please select at least one row."));
        return;
    }

    const item_codes = [...new Set(rows.map(r => r.item_code))];

    frappe.call({
        method: "generate_item.generate_item.report.requested_items_to_be_received.requested_items_to_be_received.get_last_purchase_history",
        args: { item_codes },
        callback: function (r) {
            if (r.message && r.message.length) {
                show_purchase_history_dialog(r.message);
            } else {
                frappe.msgprint(__("No purchase receipt history found."));
            }
        }
    });
}



var total_pending_qty = 0;

/* Uncheck all rows */
function unchecked_all_checkbox() {
    if (!frappe.query_report.data) return;
    
    frappe.query_report.data.forEach((row, index) => {
        let node = `.dt-row.dt-row-${index}.vrow`;
        $(node).find("[type='checkbox']").prop("checked", false);
    });
    document.querySelectorAll(".dt-row--highlight").forEach((row) => {
        row.classList.remove("dt-row--highlight");
    });
}

/* Checkbox change listener */
function listner_to_checkbox() {
    if (!frappe.query_report.data) return;
    
    frappe.query_report.data.forEach((row, index) => {
        if (row.name) {
            let node = `.dt-row.dt-row-${index}.vrow`;
            let checkbox = $(node).find("[type='checkbox']");

            checkbox.off('change').on('change', function () {
                if (this.checked) {
                    total_pending_qty += row.pending_qty || 0;
                } else {
                    total_pending_qty -= row.pending_qty || 0;
                }

                if (total_pending_qty < 0) total_pending_qty = 0;
              
            });
        }
    });
}






function group_rows_by_supplier(selected_rows) {
    const grouped = {};

    selected_rows.forEach(row => {
        if (!row.supplier) return;

        if (!grouped[row.supplier]) {
            grouped[row.supplier] = [];
        }
        grouped[row.supplier].push(row);
    });

    return grouped;
}




function create_purchase_receipt_by_supplier() {
    const selected_rows = frappe.query_report.get_checked_items() || [];

    if (!selected_rows.length) {
        frappe.msgprint(__("Please select at least one row to create purchase receipt."));
        return;
    }

    const grouped_by_supplier = group_rows_by_supplier(selected_rows);

    if (!Object.keys(grouped_by_supplier).length) {
        frappe.msgprint(__("Selected rows do not have supplier."));
        return;
    }

    const supplier_count = Object.keys(grouped_by_supplier).length;

    frappe.confirm(
        __(`This will create ${supplier_count} Purchase Receipt(s). Continue?`),
        function () {

            frappe.prompt(
                [
                    {
                        fieldname: "pr_series",
                        fieldtype: "Select",
                        label: __("PR Series"),
                        options: PO_SERIES_OPTIONS.join("\n"),
                        reqd: 1
                    },
                    {
                        fieldname: "branch",
                        fieldtype: "Link",
                        label: __("Branch"),
                        options: "Branch",
                        reqd: 1
                    }
                ],
                function (values) {
                    call_create_po(
                        grouped_by_supplier,
                        values.pr_series,
                        values.branch
                    );
                }
            );
        }
    );

    function call_create_po(grouped_by_supplier, pr_series, branch) {

        frappe.call({
            method: "generate_item.generate_item.report.requested_items_to_be_received.requested_items_to_be_received.create_purchase_receipt_by_supplier",
            args: {
                "grouped_items": grouped_by_supplier,
                "company": frappe.query_report.get_filter_value("company"),
                "pr_series": pr_series,
                "branch": branch
            },
            freeze: true,
            freeze_message: __("Creating Purchase Receipt..."),
            callback: function (r) {
                if (r.message) {
                    console.log("res----------naming series",r.message)
                    let message = __("Purchase Receipt created successfully:") + "<br><br>";

                    r.message.forEach(pr => {
                        message += `
                            <a href="/app/purchase-receipt/${pr.name}" target="_blank">
                                ${pr.name}
                            </a> (${pr.supplier})<br>
                        `;
                    });

                    frappe.msgprint({
                        title: __("Success"),
                        indicator: "green",
                        message: message
                    });

                    frappe.query_report.refresh();
                }
            }
        });
    }
}



function select_rows_with_supplier(checked) {
    total_pending_qty = 0;

    frappe.query_report.data.forEach((row, index) => {
        let node = `.dt-row.dt-row-${index}.vrow`;
        let checkbox = $(node).find("[type='checkbox']");

        checkbox.prop("checked", false);
        
    });
}

function bind_header_checkbox() {
    let header_checkbox = $(".dt-cell__content.dt-cell__content--header-0")
        .find("[type='checkbox']");

    header_checkbox.off("change").on("change", function () {
        select_rows_with_supplier(this.checked);
    });
}


$(function () {
    // footer element to show total line count below the report
    let $total_footer = $('<div class="report-footer-total-lines" style="margin-top: 8px; font-weight: bold;"></div>');
    $('.layout-main-section').append($total_footer);

    // Add action button
    frappe.query_report.page.add_action_item('Create Purchase Receipt', function () {
        create_purchase_receipt_by_supplier();
    }, __("Action"));

    // Periodically update checkbox states and listeners
    setInterval(function () {
        try {
            if (frappe.query_report && frappe.query_report.data) {
                bind_header_checkbox();     // 👈 THIS LINE IS IMPORTANT
                listner_to_checkbox();

                // update total line count below the report
                const total_rows = frappe.query_report.data.length || 0;
                $total_footer.text(__("Total Lines: {0}", [total_rows]));
            }
        } catch (error) {
            console.log(error);
        }
    }, 300);
    
    
    // Remove chart if not needed
    if (frappe.query_report.$chart) {
        frappe.query_report.$chart.remove();
    }
});


